pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/PartitionAwareRealtimeRoutingTableBuilder.java

```diff
@@ -0,0 +1,207 @@
+
+import com.linkedin.pinot.common.utils.CommonConstants;
+import com.linkedin.pinot.common.utils.LLCSegmentName;
+import com.linkedin.pinot.common.utils.SegmentName;
+import com.linkedin.pinot.common.metadata.ZKMetadataProvider;
+import com.linkedin.pinot.common.response.ServerInstance;
+import com.linkedin.pinot.transport.common.SegmentId;
+
+
+/**
+ * Partition aware routing table builder for the Kafka low level consumer
+ */
+public class PartitionAwareRealtimeRoutingTableBuilder extends AbstractPartitionAwareRoutingTableBuilder {
+
+  @Override
+  public void computeRoutingTableFromExternalView(String tableName, ExternalView externalView,
+      List<InstanceConfig> instanceConfigList) {
+
+    _numReplicas = Integer.valueOf(_tableConfig.getValidationConfig().getReplicasPerPartition());
+
+    String path = ZKMetadataProvider.constructPropertyStorePathForKafkaPartitions(tableName);
+    ZNRecord kafkaPartitionAssignment = _propertyStore.get(path, null, AccessOption.PERSISTENT);
+
+    // Compute partition id set
+    Set<Integer> partitionIds = new HashSet<>();
+    for (String segment : externalView.getPartitionSet()) {
+      SegmentId segmentId = new SegmentId(segment);
+
+      // retrieve the metadata for the segment and compute the partitionIds set
+      SegmentZKMetadata segmentZKMetadata = _segment2SegmentZkMetadataMap.get(segmentId);
+      if (segmentZKMetadata == null || segmentZKMetadata.getPartitionMetadata() == null) {
+        segmentZKMetadata = ZKMetadataProvider.getRealtimeSegmentZKMetadata(_propertyStore, tableName, segment);
+        _segment2SegmentZkMetadataMap.put(segmentId, segmentZKMetadata);
+      }
+      partitionIds.add(getPartitionId(segment));
+    }
+
+    // Compute partition to server to replica id mapping for each partition from Kafka partition mapping
+    Map<Integer, Map<ServerInstance, Integer>> serverToReplicaIdByPartitionId = new HashMap<>();
+    for (Integer partitionId : partitionIds) {
+      String partitionIdString = String.valueOf(partitionId);
+      List<String> instancesFromPartition = kafkaPartitionAssignment.getListField(partitionIdString);
+      for (int replicaId = 0; replicaId < instancesFromPartition.size(); replicaId++) {
+        String instanceName = instancesFromPartition.get(replicaId);
+        if (!serverToReplicaIdByPartitionId.containsKey(partitionId)) {
+          serverToReplicaIdByPartitionId.put(partitionId, new HashMap<ServerInstance, Integer>());
+        }
+        serverToReplicaIdByPartitionId.get(partitionId)
+            .put(ServerInstance.forInstanceName(instanceName), replicaId);
+      }
+    }
+
+    // TODO: This part of code is borrowed from `KafkaLowlevelConsumerRoutingTableBuilder`. This needs to be cleaned.
+
+    // Start of the copy code
+
+    // 1. Gather all segments and group them by Kafka partition, sorted by sequence number
+    Map<String, SortedSet<SegmentName>> sortedSegmentsByKafkaPartition = new HashMap<String, SortedSet<SegmentName>>();
```

Comment:
This part of code is borrowed from `KafkaLowLevelConsumingRoutingTableBuilder` and I will clean up this part from another pr. Anyway, the reason for using SortedSet here is to be able to pick the first segment that has `CONSUMING` state.